### PR TITLE
Developed Azure Policy & Role Bicep Modules

### PR DESCRIPTION
**Pull request by ahmadabdalla**
# Change

- Fixed bug in Upper limit in deployment script. Enforced upper limit value from parameter.
- Added Azure Policy Bicep under (Microsoft.Authorization) namespace

- policy definitions (1 main, 2 nested)
- policy initiatives (1 main, 2 nested)
- policy assignments (1 main, 3 nested - embedded role assignments)
- policy exemptions (1 main, 3 nested)
- role assignments (1 main, 3 nested)
- role Definitions (1 main, 3 nested)

Retired:
- Deploy.JSON for Policy Assignments and Role Assignments, replaced with Bicep.

```
All Modules 'Deploy.bicep' is at the `managementGroup`. This means that Service Principals using them need to perform management group deployments. All Modules support nested modules at lower level scopes covering (Management Groups, Subscriptions, Resource Groups).
```

Pending:
- Documentation for all modules
- Role Definitions Update
- Add CUAID if needed

## Type of Change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update (Wiki)

## Checklist

- [x] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [x] I did format my code

### CODE REVIEW

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 [hunk 1]
```diff
@@ -214,9 +215,7 @@ function New-ModuleDeployment {
                         $retryCount++
                     }
                 } 
-            } 
-            while ($Stoploop -eq $false -or $retryCount -eq $retryLimit) 
-
+            } until ($Stoploop -eq $true -or $retryCount -eq $retryLimit)
```
Comment (segraef) [inline .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1:218]: I leave this to @MrMCake

File: arm/Microsoft.Authorization/policyAssignments/parameters/parameters.json [hunk 2]
```diff
@@ -14,15 +14,18 @@
...
+          "value": "prod"
         }
       }
     },
     "location": {
-      "value": "westeurope"
+      "value": "australiaeast"
+    },
+    "subscriptionId": {
+      "value": "20d6fbfe-b049-471c-95af-1369d14d0d45"
```
Comment (segraef) [inline arm/Microsoft.Authorization/policyAssignments/parameters/parameters.json:28]: remove subID

File: arm/Microsoft.Authorization/policyDefinitions/parameters/parameters.json [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+          }
+        }
+      },
+      "metadata": {
+        "value": {
+          "category": "Security"
+        }
+      },
+      "subscriptionId": {
+        "value": "20d6fbfe-b049-471c-95af-1369d14d0d45"
```
Comment (segraef) [inline arm/Microsoft.Authorization/policyDefinitions/parameters/parameters.json:53]: tbr

File: arm/Microsoft.Authorization/policySetDefinitions/parameters/parameters.json [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+        ]
+      },
+      "metadata": {
+        "value": {
+          "category": "Security",
+          "version": "1"
+        }
+      },
+      "subscriptionId": {
+        "value": "20d6fbfe-b049-471c-95af-1369d14d0d45"
```
Comment (segraef) [inline arm/Microsoft.Authorization/policySetDefinitions/parameters/parameters.json:75]: tbr

File: arm/Microsoft.Authorization/roleAssignments/parameters/parametersBicep.json [hunk 5]
```diff
@@ -0,0 +1,15 @@
+{
+  "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "roleDefinitionIdOrName": {
+      "value": "Owner"
+    },
+    "principalId":{
+      "value": "9fa1a3c1-d53b-40ea-8617-ec99e51285a3"
```
Comment (segraef) [inline arm/Microsoft.Authorization/roleAssignments/parameters/parametersBicep.json:9]: obfuscate or remove

File: arm/Microsoft.Authorization/roleAssignments/parameters/parametersBicep.json [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+    },
+    "subscriptionId": {
+      "value": "20d6fbfe-b049-471c-95af-1369d14d0d45"
```
Comment (segraef) [inline arm/Microsoft.Authorization/roleAssignments/parameters/parametersBicep.json:12]: tbr

**Review by segraef — CHANGES_REQUESTED** [review]

File: arm/Microsoft.Authorization/policyAssignments/parameters/parameters.json (hunk 2 above)
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyAssignments/parameters/parameters.json:28]: I'll change to the sub ID used by the service principal

File: arm/Microsoft.Authorization/policyDefinitions/parameters/parameters.json (hunk 3 above)
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyDefinitions/parameters/parameters.json:53]: I'll change to the sub ID used by the service principal

File: arm/Microsoft.Authorization/policySetDefinitions/parameters/parameters.json (hunk 4 above)
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policySetDefinitions/parameters/parameters.json:75]: I'll change to the sub ID used by the service principal

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 [hunk 7]
```diff
@@ -77,7 +77,8 @@ function New-ModuleDeployment {
         [Parameter(Mandatory = $false)]       
         [PSCustomObject]$additionalTags,
 
-        [Parameter(Mandatory = $false)]       
+        [Parameter(Mandatory = $false)]
+        [ValidateRange(1, 3)]
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1:81]: use case: failed template, issue in upper limit input causes long time on runner/agent. Increase the number? Or remove limit at all?

File: arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_rg.bicep [hunk 8]
```diff
@@ -0,0 +1,28 @@
...
+param roleDefinitionIds array = []
+
+resource policyAssignment 'Microsoft.Authorization/policyAssignments@2020-09-01' = {
+  name: policyAssignmentName
+  location: location
+  properties: properties
+  identity: identity
+}
+
+resource roleAssignment 'Microsoft.Authorization/roleAssignments@2020-04-01-preview' = [for (roleDefinitionId, index) in roleDefinitionIds: if (!empty(roleDefinitionIds) && !empty(identity)) {
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_rg.bicep:19]: will change.

File: arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_sub.bicep [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+param roleDefinitionIds array = []
+
+resource policyAssignment 'Microsoft.Authorization/policyAssignments@2020-09-01' = {
+  name: policyAssignmentName
+  location: location
+  properties: properties
+  identity: identity
+}
+
+resource roleAssignment 'Microsoft.Authorization/roleAssignments@2020-04-01-preview' = [for (roleDefinitionId, index) in roleDefinitionIds: if (!empty(roleDefinitionIds) && !empty(identity)) {
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_sub.bicep:18]: will change

File: arm/Microsoft.Authorization/policyAssignments/deploy.bicep [hunk 10]
```diff
@@ -0,0 +1,119 @@
...
+
+@description('Optional. Location for all resources.')
+param location string = deployment().location
+
+var nonComplianceMessage_var = {
+  message: (empty(nonComplianceMessage) ? 'null' : nonComplianceMessage)
+}
+
+var var_policyAssignmentName = replace(policyAssignmentName, ' ', '-')
+var var_properties = {
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyAssignments/deploy.bicep:62]: Will align to existing converted modules, also align to bicep default (easier for users to spot). Good call out

File: arm/Microsoft.Authorization/policyExemptions/deploy.bicep [hunk 11]
```diff
@@ -0,0 +1,84 @@
...
+param managementGroupId string = ''
+
+@description('Optional. The ID of the Azure Subscription (Scope). Cannot be used with managementGroupId')
+param subscriptionId string = ''
+
+@description('Optional. The Target Scope for the Policy. The name of the resource group for the policy assignment')
+param resourceGroupName string = ''
+
+var var_policyExemptionName = toLower(replace(policyExemptionName, ' ', '-'))
+var var_policyProperties = {
```
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyExemptions/deploy.bicep:42]: Will align to existing converted modules, also align to bicep default (easier for users to spot). Good call out

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (MrMCake) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1:81]: I'd remove the limit. The pipeline will use the default value, i.e. 3, but if somebody wants to run the script e.g. locally one can provide a higher value if necessary. Whether that makes sense or not is a different question - however, I'd leave it up to the user to decide. The logic can handle any number (up to 2.147.483.647 ;) )

File: arm/Microsoft.Authorization/policyAssignments/parameters/parameters.json (hunk 2 above)
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyAssignments/parameters/parameters.json]: fixed

File: arm/Microsoft.Authorization/policyDefinitions/parameters/parameters.json (hunk 3 above)
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyDefinitions/parameters/parameters.json:53]: fixed

File: arm/Microsoft.Authorization/policySetDefinitions/parameters/parameters.json (hunk 4 above)
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policySetDefinitions/parameters/parameters.json]: fixed

File: arm/Microsoft.Authorization/roleAssignments/parameters/parametersBicep.json (hunk 5 above)
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/roleAssignments/parameters/parametersBicep.json]: Pending. Will setup a meeting to discuss.

File: .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1 (hunk 7 above)
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply .github/actions/templates/deployModule/scripts/New-ModuleDeployment.ps1]: resolved

File: arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_rg.bicep (hunk 8 above)
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_rg.bicep]: resolved

File: arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_sub.bicep (hunk 9 above)
[the remark this answers is not in the document]
Comment (ahmadabdalla) [reply arm/Microsoft.Authorization/policyAssignments/.bicep/nested_policyAssignments_sub.bicep]: resolved

**Pull request closed without merging**